### ISSUE TITLE
Copy & Paste creates a html code in a field

### ISSUE THREAD

**Issue by rmlockwood** [sillsdev/web-languageforge#136]
When I copy the Swedish word in the Word field and paste it into the Note field, it creates new lines and blank space. Even after I delete this, when I look at the field in FLEx, the note field has html elements like this: **en\<span style="font-size: 16px;float: none;">affärskvinna,\<span style="font-size: 16px;float: none;">affärskvinnor\</span>\</span>\<br/>**

**Comment by rmunn** [comment]
I suspect that this is caused by an interaction between TextAngular, the text-editing component we use in text fields, and our Send/Receive code. We recently updated the Send/Receive code to preserve < and > characters in text, because we noticed that people were using them in notes (e.g., "This word goes between the noun and the adjective, like this: \<noun> word \<adjective>"). Since Language Forge is a web page, those occurrences of \<noun> and \<adjective> between angle brackets were initially being treated as HTML tags (which were syntax errors because \<noun> is not a valid HTML tag as defined by the HTML specification). And thus, parts of users' data were getting lost, because the default behavior for websites that see incorrect HTML tags is to remove them. So we updated the Send/Receive code to preserve any < and > characters, and the text between them, so that we wouldn't lose user data.

However, the TextAngular editing component that we've been using for our input fields appears to be designed for "rich text" editing, that is, editing text where you can apply styles like bold and italic. And it does that by using HTML \<span> tags with style information in them. This is fine if all you want is to preserve the styles (bold, italic, font, etc) of the text the user typed (or pasted). But when you combine that with our recent change to preserve all text in angle brackets, it creates a problem.

We're currently discussing how to fix this, and we don't have a good answer yet. In the meantime, though, there's a workaround you can use. When you copy and paste text on a Windows computer (and Mac and Linux do something very similar), it actually stores several different "versions" of that text in the clipboard. One version is just the text with no formatting information, but another version is "text + style information". This allows you to copy and paste from a Web page into Word, and keep the bold, italic, font size, font color, etc. styles in what you paste -- since this is *usually* what people want. However, in this particular case, the "text + style information" mode of copy and pasting is causing problems due to the interaction between TextAngular and our Send/Receive feature. So one way to work around the bug is to, when you copy and paste, make sure to paste "as plain text". In Chrome (the browser I'm using right now), you can do that by right-clicking in the place where you want to paste text -- and then in the popup menu, right under "Paste (Ctrl+V)", there's a menu option "Paste as plain text (Ctrl+Shift+V)". If you're using Firefox, that menu option doesn't seem to be there, but there's another way to achieve this goal:

1. Open up Notepad.
2. Copy and paste text from your Web browser into Notepad.
3. Now select the text in Notepad, copy *that* text, and paste that into your browser.

Since Notepad is a dead-simple text editor that doesn't preserve text styles at all, the text that you paste into Notepad will be pasted in "plain-text only" mode. And then when you copy that text out of Notepad, the Windows clipboard will *not* contain the style information, and thus TextAngular won't create HTML spans to preserve the style information.

It's unfortunate that this workaround is necessary; we'll try to figure out what to do to fix the problem. But it might take us a while: we can't stop preserving < and > characters, since those really do appear in some people's data and we don't want to lose data. But we also need to use TextAngular so that data in multiple writing systems (e.g., if Greek appears in the middle of a mostly-English note) can use the proper font for each writing system. So we're caught between the Scylla of needing to preserve user data that looks like HTML, and the Charybdis of having HTML styles showing up when we don't want them.

Any progress we make on this problem, we'll report on this GitHub issue so that you can keep track of it.

**Comment by megahirt** [comment]
Robin, take a look at
https://stackoverflow.com/questions/12027137/javascript-trick-for-paste-as-plain-text-in-execcommand
 and see if we can't adapt this solution to override the paste event in
textAngular.  While I'd like to replace textAngular, it seems like it will
be a while before we can accomplish this.

**Comment by irahopkinson** [comment]
Hi Ron

I'm having trouble re-producing the problem (I was able to produce the <br/> problem but not this one). I also looked up the **Notes** in the entry for **affärskvinna** in your project but it was empty. Did you remove it or was it on another entry or another note?

Can you help with steps to re-produce or point me at some faulty data?

**Comment by rmlockwood** [comment]
Hi Ira,

 

I guess I cleaned up that entry. I’ve done the copy and paste again. I copied the text in the Word field and pasted it into the Note field and you can tell there are new lines added and extra whitespace. I’m using Chrome on Windows and just id a ctrl-C and a ctrl-V.

 

Ron

**Comment by megahirt** [comment]
I'd like to see us fix this asap.  I'm currently working on a solution that uses this code from SO (referenced previously):

$('[contenteditable]').on('paste', function(e) {
    e.preventDefault();
    var text = '';
    if (e.clipboardData || e.originalEvent.clipboardData) {
      text = (e.originalEvent || e).clipboardData.getData('text/plain');
    } else if (window.clipboardData) {
      text = window.clipboardData.getData('Text');
    }
    if (document.queryCommandSupported('insertText')) {
      document.execCommand('insertText', false, text);
    } else {
      document.execCommand('paste', false, text);
    }
});

**Comment by megahirt** [comment]
I worked on this more and unfortunately there doesn't seem to be an easy way to fix this bug without forking textAngular and modifying the source.  We plan to replace textAngular anyway, so we'll focus our efforts on replacing it (thereby getting rid of this bug) instead of mucking with more textAngular stuff.

The full story is that I was able attach an event handler for the paste event and only paste plain text.  That part worked, but I can't seem to prevent textAngular's paste handler from executing also.  TextAngular registers its paste event handler before mine (internal in it's source code) so even preventDefault(), preventPropogation() and return false don't stop TextAngular from it's own paste action.
